doc/Config.md
# Detailed Configs Guide



###### api_config.json

```json
{
  // OpenAI Base URL(must endswith "/v1")
  "api_base": "https://api.openai.com/v1",
  // OpenAI Account API_KEY
  "api_key": "",
  // ID of the model to use
  "engine": "gpt-3.5-turbo",
  // The maximum number of tokens to generate in the completion (1~4096)
  "max_tokens": 3600,
  // Control the randomness of the result (0.0~2.0), Higher values will make the output more random, while lower values will make it more focused and deterministic
  "temperature": 0.7,
  // An alternative to sampling with temperature, called nucleus sampling, where the model considers the results of the tokens with top_p probability mass
  "top_p": 1.0,
  // Positive values penalize new tokens based on whether they appear in the text so far, increasing the model's likelihood to talk about new topics (-2.0~2.0)
  "presence_penalty": 0.0,
  // Positive values penalize new tokens based on their existing frequency in the text so far, decreasing the model's likelihood to repeat the same line verbatim (-2.0~2.0)
  "frequency_penalty": 0.0,
  // How many reply to generate for each prompt
  "reply_count": 1,
  // The character played by bot by default
  "system_character": "You are ChatGPT, a large language model trained by OpenAI. Respond conversationally",

  // Optional proxy of http or https
  "proxy": ""
}
```

***1. `engine` corresponding to one of the value from：`gpt-3.5-turbo`, `gpt-3.5-turbo-0301`, `gpt-4`,  `gpt-4-0314`, `gpt-4-32k`, `gpt-4-32k-0314` 。***

***2. Please do not modify `reply_count`。***



------

###### config.json

```json
{
  // Whether to enable ChatGPT auto-reply function
  "autoReply": true,
    
  // Setting keyword to start up WebChatGPT
  "internetKey": "-o",
  // The number of searching result (1~10)
  "internetResult": 3,
    
  // Setting keyword to wake up bot in group chat
  "groupChatKey": "-c",
  // Using reply mode in group chat
  "grpReplyMode": false,
  // Origin question quote on head of answer in group chat
  "grpCitationMode": true,
  // Setting keyword to wake up bot in private chat
  "privateChatKey": "-c",
  // Using reply mode in private chat
  "prvReplyMode": true,
  // Origin question quote on head of answer in private chat
  "prvCitationMode": false,
  // Setting keyword to activate the role-playing of bot
  "characterKey": "-p",
  // Conclude and reduce token
  "conclusionKey": "-z",
  
  // Whether to enable Stable Diffusion AI drawing function (English Only)
  "stableDiffRly": true,
  // Setting keyword to wake up AI drawing function in group chat
  "groupImgKey": "-i",
  // Setting keyword to wake up AI drawing function in private chat
  "privateImgKey": "-i",
  // SSetting keyword to add negative prompt
  "negativePromptKey": "-n",
  // Whether to enable image caching (it will be cached in .cache folder)
  "isCached": true,

  // View available command help
  "helpKey": "-h",
  // Setting keyword to reset context
  "resetChatKey": "-rs",
  // Setting keyword to regenerate previous answer
  "regenerateKey": "-rg",
  // Setting keyword to roll back to previous n questions
  "rollbackKey": "-rb"
}
```

***1. The example of the WebChatGPT is: `-c -o Wandering Earth 2 story introduction`, which can trigger the online information to assist the answer.***

***2. The example of the function of accessing negative prompt sentences is: `-i skylines, 4k, concept arts -n no girl, no body`, which can use `skylines, 4k, concept arts` as a positive prompt, and `no girl, no body` as a negative prompt to stable Diffusion.***



------

###### server_config.json

```json
{
  // Setting host running locally (local only)
  "server_host": "127.0.0.1:5555",
}
```



------

###### sys_character.json

```json
{
  // "Key": "Value", (Format needs to match)
  "Linux Terminal": "I want you to act as a linux terminal. I will type commands and you will reply with what the terminal should show. I want you to only reply with the terminal output inside one unique code block, and nothing else. do not write explanations. do not type commands unless I instruct you to do so. when i need to tell you something in english, i will do so by putting text inside curly brackets {like this}. my first command is pwd",
  .
  .
  .
}
```

This file is used to store all the custom characters that the bot can play (note that the custom character name does not allow empty characters on both sides, and there can be empty characters on both sides of the character name when activated). 

Notice! If you want to activate it, you need to use a custom order, and the activated character exists in the configuration file. For example, according to the default configuration, use the command "-pLinux Terminal" or "-p Linux Terminal  	" in chat to activate the value corresponding to `Linux Terminal` as the role played by bot.



doc/Config_ZH.md
### 配置详细指引



###### api_config.json

```json
{
  // OpenAI Base URL（也可以是任意兼容 OpenAI 的 API 服务地址）
  "api_base": "https://api.openai.com/v1",
  // OpenAI 账户构建的 API_KEY
  "api_key": "",
  // 需要的模型名称
  "engine": "gpt-3.5-turbo",
  // 最大可接受提问 token（1~4096）
  "max_tokens": 3600,
  // 控制结果随机性（0.0~2.0），0.0 表示结果固定，2.0 表示结果随机
  "temperature": 0.7,
  // 使用温度采样的替代方法称为核心采样，可以采取默认
  "top_p": 1.0,
  // 主题贴合度惩罚（-2.0~2.0），正值会根据到目前为止是否出现在文本中来惩罚新标记，从而增加模型谈论新主题的可能性
  "presence_penalty": 0.0,
  // 频率惩罚（-2.0~2.0），正值会根据新标记在文本中的现有频率对其进行惩罚，从而降低模型逐字重复同一行的可能性
  "frequency_penalty": 0.0,
  // 获得的回答数
  "reply_count": 1,
  // 机器人默认扮演的角色
  "system_character": "You are ChatGPT, a large language model trained by OpenAI. Respond conversationally",

  // 可选填写 http 或 https 代理
  "proxy": ""
}
```

***1. `engine` 对应的值需要为其中之一：`gpt-3.5-turbo`, `gpt-3.5-turbo-0301`, `gpt-4`,  `gpt-4-0314`, `gpt-4-32k`, `gpt-4-32k-0314` 。***

***2. 请不要更改 `reply_count`。***



------

###### config.json

```json
{
  // 是否开启 ChatGPT 自动回复
  "autoReply": true,
    
  // 在回复时，查询联网信息 (WebChatGPT) 辅助回答
  "internetKey": "-o",
  // 联网信息条数 (1~10)
  "internetResult": 3,
  
  // 在群聊中设置唤醒机器人关键词
  "groupChatKey": "-c",
  // 在群聊中响应回复
  "grpReplyMode": true,
  // 在群聊回答前添加源问题格式
  "grpCitationMode": true,
  // 在私聊中设置唤醒机器人关键词
  "privateChatKey": "-c",
  // 在私聊中响应回复
  "prvReplyMode": true,
  // 在私聊回答前添加源问题格式
  "prvCitationMode": false,
  // 激活机器人扮演角色关键词
  "characterKey": "-p",
  // 总结对话并缩减 token
  "conclusionKey": "-z",

  // 是否开启 Stable Diffusion 图片回复（仅英语）
  "stableDiffRly": true,
  // 在群聊中设置唤醒 AI画图功能关键词
  "groupImgKey": "-i",
  // 在私聊中设置唤醒 AI画图功能关键词
  "privateImgKey": "-i",
  // 接入负语句关键词
  "negativePromptKey": "-n",
  // 是否开启图片缓存（开启后会在 .cache 文件夹中缓存）
  "isCached": true,

  // 查看可用命令帮助
  "helpKey": "-h",
  // 设置重置上下文关键词
  "resetChatKey": "-rs",
  // 设置重新生成答案关键词
  "regenerateKey": "-rg",
  // 设置回滚到以前的n个问题关键词
  "rollbackKey": "-rb"
}
```

***1. 联网查询信息功能（WebChatGPT）示例为：`-c -o 流浪地球2故事简介` 即可触发联网信息辅助问答。***

***2. 接入负反馈语句关键词功能示例为：`-i skylines, 4k, concept arts -n no girl, no body` 即可将 `skylines, 4k, concept arts` 作为正提示， `no girl, no body` 作为负提示输入给 stable Diffusion。***



------

###### server_config.json

```json
{
  // 本地host运行地址（仅本地）
  "server_host": "127.0.0.1:5555",
}
```



------

###### sys_character.json

```json
{
  // "键": "值", (格式需匹配)
  "Linux Terminal": "I want you to act as a linux terminal. I will type commands and you will reply with what the terminal should show. I want you to only reply with the terminal output inside one unique code block, and nothing else. do not write explanations. do not type commands unless I instruct you to do so. when i need to tell you something in english, i will do so by putting text inside curly brackets {like this}. my first command is pwd",
  .
  .
  .
}
```

这个文件用于存放自定义机器人可扮演的所有角色（注意自定义角色名不允许两侧包含空字符，激活时角色名两侧可以有空字符）。

注意！如果想要激活它，需要用自定义指令，且激活的角色在该配置文件中存在。例如按照默认配置，在聊天框中使用命令 "-pLinux Terminal" 或 "-p  Linux Terminal   	"，均可激活 `Linux Terminal` 对应的值作为机器人扮演的角色。



shared/config.py
# -*- coding: utf-8 -*-

import json

# api model check
with open('.config/api_config.json', encoding='utf-8') as f:
    api_config = json.load(f)

if api_config['engine'] not in [
    'gpt-3.5-turbo',
    'gpt-3.5-turbo-0301',
    'gpt-3.5-turbo-16k',
    'gpt-3.5-turbo-0613',
    'gpt-3.5-turbo-16k-0613',
    'gpt-4',
    'gpt-4-0314',
    'gpt-4-32k',
    'gpt-4-32k-0314',
    'gpt-4-0613',
    'gpt-4-32k-0613',
]:
    error = NotImplementedError('Unsupported engine {self.engine}')
    raise error

# general config
with open('.config/config.json', encoding='utf-8') as f:
    config = json.load(f)
f.close()

autoReply = config['autoReply']

internetKey = config['internetKey']
internetResult = config['internetResult']

groupChatKey = config['groupChatKey']
grpReplyMode = config['grpReplyMode']
grpCitationMode = config['grpCitationMode']
privateChatKey = config['privateChatKey']
prvReplyMode = config['prvReplyMode']
prvCitationMode = config['prvCitationMode']
characterKey = config['characterKey']
conclusionKey = config['conclusionKey']

stableDiffRly = config['stableDiffRly']
groupImgKey = config['groupImgKey']
privateImgKey = config['privateImgKey']
negativePromptKey = config['negativePromptKey']
isCached = config['isCached']

helpKey = config['helpKey']
resetChatKey = config['resetChatKey']
regenerateKey = config['regenerateKey']
rollbackKey = config['rollbackKey']